### ISSUE TITLE
Move most file handling logic to server

### ISSUE THREAD

**Issue by gvidon** [okfn/datapackagist#171]
```tabularfilemanager.fromURL()``` could perfectly replace routine of ```.getDataSource()``` https://github.com/okfn/datapackagist/blob/master/datapackagist/src/scripts/components/ui/resource-editor.js used for fetching and parsing resources files from remote URL.

It will also help to avoid hanging on large CSVs, as there is an option defining max part of the CSV to be processed and infered into schema.

**Comment by pwalsh** [comment]
@gvidon ok, great. 

So, we should provide a distinct service on our service for CSV/Excel, that uses tabularfilemanager code (distinct from the service that will take json, for example).

But what about uploads? We'd want to work consistently with URL and upload. I know that tabularfilemanager should do this too, right? If it simplifies things, you can take the current tabularfilemanager code, and strip out the stuff that is there for browser compatibility, so that we have a good server-side solution with no compromises. Once we get it working in the codebase, we can maybe package it up for npm. WDYT?

**Comment by gvidon** [comment]
@pwalsh yes, ```tabularfilemanager``` can process both local files and URLs. So, what you mean is backend view that takes either URL from GET (or from url param) or file data from POST. As a result this view should response with infered schema and parsed CSV (and some other properties returned by ```tabularfilemanager```)?

And do we keep local CORS proxy for json?

**Comment by pwalsh** [comment]
Yes to all above.

It simplifies the client and puts most processing into the server.

Does this sound good to you? Any problems?

**Comment by gvidon** [comment]
@pwalsh it sounds good in terms of an overall consistency. The only thing that is not absolutely clear to me is how to refine generic upload dialog according to new conditions and still keep it as generic solution — https://github.com/okfn/datapackagist/blob/master/datapackagist/src/scripts/components/ui/upload-dialog.js. But it is a minor problem.

**Comment by pwalsh** [comment]
@gvidon great.

**Comment by gvidon** [comment]
```tabularfilemanager``` validates resources over goodtables API. So, https://github.com/okfn/datapackagist/blob/master/datapackagist/src/scripts/components/ui/validationresult.js should be refined with this in mind.

- [ ] Replace validation routine with reading of resources attributes which were returned by ```tabularfilemanager``` from server side.

@pwalsh should I move this to distinct issue?

**Comment by pwalsh** [comment]
@gvidon renaming, and adding a specific list of tasks (and, please comment on stuff that I miss in the list)

**Comment by pwalsh** [comment]
# Description

We currently do all sorts of file processing purely client side (datapackage.json files and data files). For various reasons (processing speed, CORS), we want to move **some** logic to the server.

# Tasks

* [ ] Include a "next iteration" of our `tabularfilemanager` lib, that runs on the server. If it cleans up the API, then, remove aspects of this code that were originally written for browser + server compat.
* [ ] Expose a suite of services `service` (service > controller + route, service can be unit tested, controller + route can have functional tests) for the following:
  * [ ] When the client receives a `datapackage.json` for upload or from URL, get the server to handle this, and it will return the file to the client (i.e.: Avoid CORS via a self-hosted proxy)
  * [ ] When the client receives a tabular data file (i.e.: CSV or Excel), costume this via an endpoint that wraps `tabularfilemanager`. The settings for TFM should be exposed as params on the endpoint (e.g.: `infer`, `validate`)
  * [ ] Do we need any special additional service for our `datapackage-from-remote` routines, or, will it be covered by the above two? 
* [ ] Make client adjustments as required to work with the new data flow (example, as noted [here](https://github.com/okfn/datapackagist/issues/171#issuecomment-147768352))
* [ ] Write unit tests for the services
* [ ] Write functional tests for the API endpoints

**Comment by gvidon** [comment]
@pwalsh what I don't like about server side solution
* as we discussed today — when user uploads large CSVs from local drive we will have to cut it locally and only after this send to server, as in other case large CSVs upload will create serious lag. This is like architecture design issue, because ```tabularfilemanager``` has a feature for cutting large data files — so, we will have to copy this feature on client side.
* when user uploads ```datapackage.json``` from local drive why do we need to send it to the server? Probably I just misunderstood you?
* having special server side service for ```datapackage-from-remote``` would be useful when user gets ```datapackage.json``` with resources which schemas are not yet defined and we need to infer schemas for each resource. But I'm not sure if we need this feature.

**Comment by pwalsh** [comment]
@gvidon 

* on the server, we can read the CSV as a stream. I don't think we need to cut in client. Large files take time.... we can't remove that issue. Most files, however, are not large.
* don't need to send to server for local file of datapackage.json, for any reason, except to have a *consistent API*. I prefer we don't have two different code paths.

**Comment by pwalsh** [comment]
@gvidon as we agreed now on Skype - the handling of the datapackage.json can be done any way you prefer.

**Comment by richhall** [comment]
@pwalsh to clarify
1. will this resolve current CORS issue with processing CSV files referenced in a data package?
2. During server-side processing will the UI hang or get some indication of progress?

Apologies if these are already answered!

**Comment by gvidon** [comment]
@pwalsh 
* IE9 doesn't support ```FormData```. I'm able to polyfill it with https://github.com/francois2metz/html5-formdata
* IE9 ignores ```Access-Control-Allow``` headers, so cross domain requests are almost impossible there. ```superagent``` had a solution, but then they removed it https://github.com/visionmedia/superagent/issues/409 We use ```superagent``` everywhere for http requests, custom solution will require us to rewrite much code.

Cross domain requests used to fetch registry profiles list. We can move this to backend if IE9 support is critical.

I'm not sure that registry list fetching is the only problem with IE9, possibly there will be others, need to test it.

**Comment by gvidon** [comment]
Firefox 39 is ok

**Comment by gvidon** [comment]
@pwalsh should implement registry profiles fetching on backend?

How about ```datapackage-from-remote``` — do we decide to place it on backend too, and do resources validation and infer once data package received from remote source?

**Comment by gvidon** [comment]
Generic upload solution is not suitable for streamed upload or sending file in chunks, as http multipart spec requires file content to be in single request.

Existing solutions like https://github.com/balderdashy/skipper from what I understand still uploads the whole http request data (request body) to the server and then stream it into different sources, not creating temporary file. It doesn't eliminate upload lag.

What should eliminate upload lag is splitting resource file into chunks on the client and use http request ```Conten-range``` header thttp://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.16 to send data in chunks.

But probably much simpler would be to just truncate blob from file input on the client, and send already truncated data to the server

Here it is how client side truncation is done.
```javascript
// .slice() returns Blob which will be sent to the backend
file.slice(0, _.min([file.size, config.maxCSVSize]))
```

**Comment by pwalsh** [comment]
@gvidon 

* For IE9, CORS, etc: https://github.com/matthew-andrews/isomorphic-fetch
* For large files: let's not overthink this. Large files take time to process, it is a fact. The point of moving to server is to prevent lag in client, not to magically solve processing of large files. With server managing this, the client has less to do. I see no need whatsoever to start truncating files on client before sending to the file manager backend.

**Comment by gvidon** [comment]
@pwalsh I'm not worrying about processing large files. I'm worrying about uploading large files to the server from user browser.

**Comment by gvidon** [comment]
@pwalsh we will have to move from superagent to fetch in datapackage-from-remote, datapackage-registry and datapackagist itself. All these packages are used in browser currently.

Is it acceptable?

**Comment by pwalsh** [comment]
After lots of investigation, closing in favour of new issues to deal with this issue differently.
